chainer/_link.py

```diff
@@ -0,0 +1,271 @@
+                used.
+
+        Returns: self
+
+        """
+        cuda.check_cuda_available()
+        self._device = None
+        cuda_device = cuda._get_device_or_current(device)
+        device = chainer.backends.cuda.GpuDevice(cuda_device)
+        self.visit_device_objects(_ToDeviceVisitor(
+            device,
+            skip_between_cupy_devices=True))
+        return self
+
+    def to_intel64(self):
+        # type: () -> 'DeviceObjectsVisitableMixin'
+        """Copies parameter variables and persistent values to CPU."""
+        intel64.check_ideep_available()
+        return self.to_device(intel64)
+
+    def to_chainerx(self):
+        """Converts parameter variables and persistent values to ChainerX \
+without any copy.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to ChainerX, the link implementation must
+        override this method to do so.
+
+        Returns: self
+        """  # NOQA
+        if not chainerx.is_available():
+            raise RuntimeError('ChainerX is not available.')
+
+        if self.xp is chainerx:
+            return self
+
+        self._device = None
+        self.visit_device_objects(_ToChainerxVisitor())
+        return self
+
+    def from_chainerx(self):
+        """Converts parameter variables and persistent values from ChainerX \
+to NumPy/CuPy devices without any copy."""
+        if isinstance(self._device, backend.ChainerxDevice):
+            self._device = self._device.fallback_device
+
+        self._device = None
+        self.visit_device_objects(_FromChainerxVisitor())
+        return self
+
+    def to_device(self, device):
+        # type: (types.DeviceSpec) -> 'DeviceObjectsVisitableMixin'
+        """Copies parameter variables and persistent values to the specified \
+device.
+
+        This method does not handle non-registered attributes. If some of such
+        attributes must be copied to the device, the link implementation must
+        override this method to do so.
+
+        Args:
+            device: Target device specifier. See
+                :func:`~chainer.get_device` for available values.
+
+        Returns: self
+
+        """  # NOQA
+        device = chainer.get_device(device)
+        self._device = None
+        self.visit_device_objects(_ToDeviceVisitor(device))
+        return self
+
+
+class DeviceObjectsVisitor(object):
+    def visit_visitable(self, visitable, visitor):
+        assert isinstance(visitable, DeviceObjectsVisitableMixin)
+        visitable.visit_device_objects(visitor)
+
+    def visit_array(self, arr):
+        """Processes an array and returns a new one"""
+        raise NotImplementedError()
+
+    def visit_param(self, param):
+        """Processes a parameter"""
+        raise NotImplementedError()
+
+    def visit_link(self, link):
+        """Processes a link"""
+        raise NotImplementedError()
+
+
+class _GetSoleDeviceVisitor(DeviceObjectsVisitor):
```

Comment:
Is this visitor needed to be dedicated to getting the "sole" device? Can we generalize it to a visitor that collects all devices used in the object and let the user of this visitor determine if it is "sole" from the length of the result?

Reply:
Yes, we can, but I think there's no need for generalization at the moment because it's a private class.

Reply:
Hmm, ok. I just thought the generalized one is easier to understand for me.